File: src/lexer.rs
use crate::{
    logger::{self, log},
    token::{Token, TokenKind},
};
use std::{collections::HashMap, ops::Deref, time::Instant};

/// Lexer struct that holds the input string, current lexer position and character
pub struct Lexer {
    input: String,
    cur_char: char,
    pos: usize,
    line: usize,
    keywords: HashMap<String, TokenKind>,
}

impl Lexer {
    /// instantiate a new Lexer
    pub fn new(input: String) -> Lexer {
        Lexer {
            input,
            cur_char: '\0',
            pos: 0,
            line: 0,
            keywords: HashMap::from([
                ("and".to_string(), TokenKind::AND),
                ("or".to_string(), TokenKind::OR),
                ("fun".to_string(), TokenKind::FUN),
                ("for".to_string(), TokenKind::FOR),
                ("if".to_string(), TokenKind::IF),
                ("else".to_string(), TokenKind::ELSE),
                ("true".to_string(), TokenKind::TRUE),
                ("false".to_string(), TokenKind::FALSE),
                ("nil".to_string(), TokenKind::NIL),
                ("return".to_string(), TokenKind::RETURN),
                ("var".to_string(), TokenKind::VAR),
                ("while".to_string(), TokenKind::WHILE),
                ("class".to_string(), TokenKind::CLASS),
                ("super".to_string(), TokenKind::SUPER),
                ("this".to_string(), TokenKind::THIS),
                ("print".to_string(), TokenKind::PRINT),
            ]),
        }
    }

    fn error(&self, msg: &str, description: &str, lexeme: &str) {
        log().syntax_error(&format!(
            "{} at pos '{}' of '{}' on line {}:\n",
            msg, self.pos, self.input, self.line
        ));

        let line = if self.line < 9 {
            format!("0{}", self.line)
        } else {
            format!("{}", self.line)
        };

        let up_arrows = if lexeme.len() > 1 {
            format!(" {}", "^".repeat(lexeme.len()))
        } else {
            "^".to_string()
        };

        println!(
            "{}{} |{} {}\n{}{} {}{}\n\n{}\n",
            logger::COLOR_BLUE,
            line,
            logger::COLOR_RESET,
            self.input,
            logger::COLOR_RED,
            " ".repeat((self.pos - lexeme.len()) + 5) + &up_arrows,
            msg,
            logger::COLOR_RESET,
            description
        );
    }

    fn at_end(&self) -> bool {
        self.pos >= self.input.len() || self.cur_char == '\0'
    }

    fn advance(&mut self) {
        if self.pos + 1 < self.input.len() {
            self.pos += 1;
            self.cur_char = self
                .input
                .chars()
                .nth(self.pos)
                .expect("error while retrieving character from input");
        } else {
            self.cur_char = '\0';
        }
    }

    fn number(&mut self) -> (TokenKind, String) {
        let mut is_float = false;
        let mut result = String::new();
        while self.cur_char.is_digit(10) {
            result.push(self.cur_char);
            self.advance();
        }
        if self.cur_char == '.' && self.peek().is_digit(10) {
            is_float = true;

            result.push(self.cur_char);

            // skip the '.'
            self.advance();

            while self.cur_char.is_digit(10) {
                result.push(self.cur_char);
                self.advance();
            }
        }
        if is_float {
            (
                TokenKind::FLOAT(result.parse::<f64>().expect("failed to parse float")),
                result,
            )
        } else {
            (
                TokenKind::INTEGER(result.parse::<usize>().expect("failed to parse integer")),
                result,
            )
        }
    }

    fn string(&mut self) -> (TokenKind, String) {
        let mut result = String::new();

        // skip the "
        self.advance();

        while self.cur_char != '"' && !self.at_end() {
            result.push(self.cur_char);
            self.advance();
        }

        if self.at_end() {
            self.error(
                "unterminated string",
                "string literals must be terminated with '\"'",
                &result,
            );
            return (TokenKind::UNKNOWN(), result);
        }

        (TokenKind::STRING(result.clone()), result)
    }

    fn is_alpha(&self) -> bool {
        self.cur_char.is_alphabetic() || self.cur_char == '_'
    }

    fn is_alphanumeric(&self) -> bool {
        self.is_alpha() || self.cur_char.is_digit(10)
    }

    fn identifier(&mut self) -> (TokenKind, String) {
        let mut result = String::new();
        while self.is_alphanumeric() {
            result.push(self.cur_char);
            self.advance();
        }

        let mut token_kind = TokenKind::IDENTIFIER(result.clone());
        let res = self.keywords.get(&result);

        token_kind = match res {
            Some(kind) => kind.clone(),
            None => token_kind,
        };

        (token_kind, result)
    }

    fn peek(&self) -> char {
        if self.at_end() {
            '\0'
        } else {
            self.input.chars().nth(self.pos + 1).expect("Couldn't peek")
        }
    }

    fn peek_equals(&self, c: char) -> bool {
        if self.at_end() {
            return false;
        }

        match self.input.chars().nth(self.pos + 1) {
            Some(next_char) => next_char == c,
            None => false,
        }
    }

    pub fn lex(&mut self) -> Vec<Token> {
        let start_time = Instant::now();
        let mut vectors = vec![];
        if self.input.is_empty() {
            return vectors;
        }

        self.cur_char = self
            .input
            .chars()
            .nth(self.pos)
            .expect("error while retrieving character from input");

        while self.cur_char != '\0' {
            let token_kind: TokenKind;
            let mut literal: String = String::new();
            let cur_char_num = self.cur_char.is_digit(10);
            match self.cur_char {
                ' ' | '\t' | '\r' => {
                    self.advance();
                    continue;
                }
                '\n' => {
                    self.line += 1;
                    self.pos = 0;
                    continue;
                }
                '\0' => token_kind = TokenKind::EOF,
                '"' => {
                    let (t_kind, t_literal) = self.string();
                    token_kind = t_kind;
                    literal = t_literal;
                    if token_kind == TokenKind::UNKNOWN() {
                        return vec![];
                    }
                }
                '+' => token_kind = TokenKind::PLUS,
                '-' => token_kind = TokenKind::MINUS,
                ',' => token_kind = TokenKind::COMMA,
                ';' => token_kind = TokenKind::SEMICOLON,
                '/' => {
                    if self.peek_equals('/') {
                        while self.cur_char != '\n' && !self.at_end() {
                            self.advance();
                        }
                        continue;
                    } else {
                        token_kind = TokenKind::SLASH;
                    }
                }
                '.' => token_kind = TokenKind::DOT,
                '(' => token_kind = TokenKind::OPENPAREN,
                ')' => token_kind = TokenKind::CLOSEPAREN,
                '{' => token_kind = TokenKind::RIGHTBRACE,
                '}' => token_kind = TokenKind::LEFTBRACE,
                '%' => token_kind = TokenKind::MOD,
                '=' => token_kind = TokenKind::EQUAL,
                '<' => {
                    token_kind = if self.peek_equals('=') {
                        self.advance();
                        TokenKind::LESSTHANEQUAL
                    } else {
                        TokenKind::LESSTHAN
                    }
                }
                '>' => {
                    token_kind = if self.peek_equals('=') {
                        self.advance();
                        TokenKind::GREATERTHANEQUAL
                    } else {
                        TokenKind::GREATERTHAN
                    }
                }
                '!' => {
                    token_kind = if self.peek_equals('=') {
                        self.advance();
                        TokenKind::BANGEQUAL
                    } else {
                        TokenKind::BANG
                    }
                }
                '*' => {
                    token_kind = if self.peek_equals('*') {
                        self.advance();
                        TokenKind::EXPONENT
                    } else {
                        TokenKind::ASTERISK
                    }
                }
                _ => {
                    if cur_char_num {
                        let (number, new_literal) = self.number();
                        literal = new_literal;
                        token_kind = number;
                    } else if self.is_alpha() {
                        let (identifier, new_literal) = self.identifier();
                        literal = new_literal;
                        token_kind = identifier;
                    } else {
                        self.error(
                            &format!("unexpected character '{}'", self.cur_char),
                            "expected a valid character, such as a number, operator, or parenthesis",
                            &literal,
                        );
                        return vec![];
                    }
                }
            }

            let pos = match self.pos {
                0 => 0,
                x if x < literal.len() => 0,
                _ => self.pos - literal.len(),
            };

            let t = Token {
                pos,
                kind: token_kind,
                literal,
                line: self.line,
            };

            vectors.push(t);

            if !cur_char_num {
                self.advance();
            }
        }

        vectors.push(Token {
            pos: self.pos,
            kind: TokenKind::EOF,
            literal: String::new(),
            line: self.line,
        });

        log().debug(&format!(
            "parsed: {} token{}, took {}µs",
            vectors.len(),
            if vectors.len() == 1 { "" } else { "s" },
            start_time.elapsed().as_micros()
        ));

        vectors
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexer_symbols() {
        let mut lexer = Lexer::new(
            "5192 * 17.9 % + - / ** (  ) ! != = . < > <= >= \"this is a string\"".to_string(),
        );
        let tokens = lexer.lex();
        assert_eq!(tokens[0].kind, TokenKind::INTEGER(5192));
        assert_eq!(tokens[1].kind, TokenKind::ASTERISK);
        assert_eq!(tokens[2].kind, TokenKind::FLOAT(17.9));
        assert_eq!(tokens[3].kind, TokenKind::MOD);
        assert_eq!(tokens[4].kind, TokenKind::PLUS);
        assert_eq!(tokens[5].kind, TokenKind::MINUS);
        assert_eq!(tokens[6].kind, TokenKind::SLASH);
        assert_eq!(tokens[7].kind, TokenKind::EXPONENT);
        assert_eq!(tokens[8].kind, TokenKind::OPENPAREN);
        assert_eq!(tokens[9].kind, TokenKind::CLOSEPAREN);
        assert_eq!(tokens[10].kind, TokenKind::BANG);
        assert_eq!(tokens[11].kind, TokenKind::BANGEQUAL);
        assert_eq!(tokens[12].kind, TokenKind::EQUAL);
        assert_eq!(tokens[13].kind, TokenKind::DOT);
        assert_eq!(tokens[14].kind, TokenKind::LESSTHAN);
        assert_eq!(tokens[15].kind, TokenKind::GREATERTHAN);
        assert_eq!(tokens[16].kind, TokenKind::LESSTHANEQUAL);
        assert_eq!(tokens[17].kind, TokenKind::GREATERTHANEQUAL);
        assert_eq!(
            tokens[18].kind,
            TokenKind::STRING("this is a string".to_string())
        );
    }

    #[test]
    fn test_lexer_skip_whitespace() {
        let mut lexer = Lexer::new(" \n\t".to_string());
        let tokens = lexer.lex();
        assert_eq!(tokens.len(), 0);
    }

    #[test]
    fn test_lexer_skip_unknown() {
        let mut lexer = Lexer::new("unknown".to_string());
        let tokens = lexer.lex();
        for token in tokens {
            assert_eq!(token.kind, TokenKind::UNKNOWN());
        }
    }

    #[test]
    fn test_lexer_string() {
        let mut lexer =
            Lexer::new("\"this is the first string\" \"second string hmmmmm\"".to_string());
        let tokens = lexer.lex();
        assert_eq!(
            tokens[0].kind,
            TokenKind::STRING("this is the first string".to_string())
        );
        assert_eq!(
            tokens[1].kind,
            TokenKind::STRING("second string hmmmmm".to_string())
        );
    }
}

File: lexer/lexer.go
package lexer

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"
	"unicode"

	"github.com/xnacly/tiny-interpreter/logger"
	"github.com/xnacly/tiny-interpreter/token"
)

type Lexer struct {
	input       []rune // the input to lex
	position    int    // current position in input (points to current char)
	currentChar rune   // current character
	line        int    // position in the file
	line_pos    int    // position in the current line
}

// creates a new instance of Lexer and returns it
func NewLexer(input string) *Lexer {
	l := &Lexer{input: []rune(input), line: 1, line_pos: 0, position: 0}
	if len(input) > 0 {
		l.currentChar = l.input[l.position]
	} else {
		l.currentChar = 0
	}
	return l
}

// advances the lexer by one character
func (l *Lexer) advance() {
	if l.position+1 < len(l.input) {
		l.position++
		l.line_pos++
		l.currentChar = l.input[l.position]
	} else {
		l.currentChar = 0
	}
}

// returns true if the current character is EOF or the length of the input is reached
func (l *Lexer) atEnd() bool {
	return l.position >= len(l.input) || l.currentChar == 0
}

// returns the next character without advancing the lexer
func (l *Lexer) peek() rune {
	if l.atEnd() || l.position+1 >= len(l.input) {
		return 0
	} else {
		return l.input[l.position+1]
	}
}

// returns true if the next character is equal to c
func (l *Lexer) peekEquals(c rune) bool {
	return c == l.peek()
}

// returns true if the current character is in the alphabet
func (l *Lexer) isAlpha() bool {
	return unicode.IsLetter(l.currentChar)
}

// returns true if the current character is in the alphabet or is a digit
func (l *Lexer) isAlphaNumeric() bool {
	return l.isAlpha() || unicode.IsDigit(l.currentChar)
}

// advances the lexer until the current character is a closing quote, returns the string, throws an error if the end of the input is reached before closing the string
func (l *Lexer) string() (string, error) {
	res := ""

	l.advance()

	for l.currentChar != '"' && !l.atEnd() {
		res += string(l.currentChar)
		l.advance()
	}

	if l.atEnd() {
		return res, errors.New("unterminated string")
	} else {
		return res, nil
	}
}

// advances the lexer until the current character is not a digit, returns the digit
func (l *Lexer) number() (string, float64, error) {
	res := ""
	for unicode.IsDigit(l.currentChar) {
		res += string(l.currentChar)
		l.advance()
	}
	if l.currentChar == '.' && unicode.IsDigit(l.peek()) {
		res += string(l.currentChar)

		l.advance()

		for unicode.IsDigit(l.currentChar) {
			res += string(l.currentChar)
			l.advance()
		}
	}
	integer, err := strconv.ParseFloat(res, 64)
	return res, integer, err
}

// advances the lexer until the current character is a whitespace, returns the identifier
func (l *Lexer) identifier() (token.TokenType, string) {
	res := ""
	for l.isAlphaNumeric() {
		res += string(l.currentChar)
		l.advance()
	}

	if val, ok := token.KEYWORDS[res]; ok {
		return val, res
	} else {
		return val, res
	}
}

func (l *Lexer) error(message string, description string, lexeme string) {
	errorLine := strings.Split(string(l.input), "\n")[l.line-1]

	linePosFormatted := ""
	if l.line < 9 {
		linePosFormatted = "0" + fmt.Sprint(l.line)
	} else {
		linePosFormatted = fmt.Sprint(l.line)
	}

	fmt.Printf("%serror%s: %s at pos %d of '%s' on line %d\n\n", logger.ANSI_RED, logger.ANSI_RESET, message, l.line_pos, errorLine, l.line)

	upArrows := "^"
	if len(lexeme) > 1 {
		upArrows = strings.Repeat(upArrows, len(lexeme))
	}

	spaces := ""
	if l.line_pos < len(lexeme) {
		spaces = strings.Repeat(" ", l.line_pos)
	} else {
		spaces = strings.Repeat(" ", (l.line_pos-len(lexeme))+6) + upArrows
	}

	fmt.Printf("%s%s |%s %s\n%s%s %s%s\n\n%s\n",
		logger.ANSI_BLUE,
		linePosFormatted,
		logger.ANSI_RESET,
		errorLine,
		logger.ANSI_RED,
		spaces,
		message,
		logger.ANSI_RESET,
		description,
	)
}

// lexes the input and returns a slice of tokens
func (l *Lexer) Lex() []token.Token {
	startTime := time.Now()
	tokens := []token.Token{}
	for !l.atEnd() {
		var kind token.TokenType = token.UNKNOWN
		var literal string = ""
		var value any
		switch l.currentChar {
		case ' ', '\t', '\r':
			l.advance()
			continue
		case '\n':
			l.line++
			l.line_pos = -1
			l.advance()
			continue
		case 0:
			kind = token.EOF
		case '"':
			lit, err := l.string()
			if err != nil {
				l.error("Unterminated string", "string literals must be terminated with '\"'", lit)
				return []token.Token{}
			}
			literal = lit
			value = lit
			kind = token.STRING
		case '+':
			kind = token.PLUS
		case '-':
			kind = token.MINUS
		case ',':
			kind = token.COMMA
		case ';':
			kind = token.SEMICOLON
		case '/':
			if l.peekEquals('/') {
				for l.currentChar != '\n' && !l.atEnd() {
					l.advance()
				}
				continue
			} else {
				kind = token.SLASH
			}
		case '.':
			kind = token.DOT
		case '(':
			kind = token.OPENPAREN
		case ')':
			kind = token.CLOSEPAREN
		case '{':
			kind = token.LEFTBRACE
		case '}':
			kind = token.RIGHTBRACE
		case '%':
			kind = token.MOD
		case '=':
			kind = token.EQUAL
		case '<':
			if l.peekEquals('=') {
				l.advance()
				kind = token.LESSTHANEQUAL
			} else {
				kind = token.LESSTHAN
			}
		case '>':
			if l.peekEquals('=') {
				l.advance()
				kind = token.GREATERTHANEQUAL
			} else {
				kind = token.GREATERTHAN
			}
		case '!':
			if l.peekEquals('=') {
				l.advance()
				kind = token.BANGEQUAL
			} else {
				kind = token.BANG
			}
		case '*':
			if l.peekEquals('*') {
				l.advance()
				kind = token.EXPONENT
			} else {
				kind = token.ASTERISK
			}
		default:
			if unicode.IsDigit(l.currentChar) {
				l, v, err := l.number()
				if err != nil {
					logger.LError("couldn't parse integer")
				}
				literal = l
				value = v
				kind = token.INTEGER
			} else if l.isAlpha() {
				t, l := l.identifier()
				literal = l
				kind = t
			} else {
				l.error("Unexpected character '"+string(l.currentChar)+"'", "expected a valid character, such as a number, operator, or parenthesis", string(l.currentChar))
				return []token.Token{}
			}
		}

		pos := l.line_pos
		if pos == 0 {
			pos = 0
		} else if pos < len(literal) {
			pos = 0
		} else {
			pos = l.line_pos - len(literal)
		}

		tokens = append(tokens, token.Token{
			Type:    kind,
			Literal: literal,
			Value:   value,
			Pos:     pos,
			Line:    l.line,
		})
		if kind != token.INTEGER {
			l.advance()
		}
	}
	tokens = append(tokens, token.Token{
		Type:    token.EOF,
		Pos:     l.line_pos + 1,
		Line:    l.line,
		Literal: "",
	})
	logger.LInfo(fmt.Sprintf("(f)lexed: %d tokens, took %s", len(tokens), time.Since(startTime).String()))
	return tokens
}
